Add Dockerfile

sbellem opened the pull request:
Just saw and commented in #523, and thought that perhaps adding a `Dockerfile` could be useful. I can adapt the `Dockerfile` as needed too.

mkskeller commented:
Thank you for submitting this. What is the reasoning for activating a raft of debugging flags by default? I could imagine that some protocols will produce somewhat garbled output when `DEBUG_MAC` is activated for example.

sbellem commented:
> What is the reasoning for activating a raft of debugging flags by default? I could imagine that some protocols will produce somewhat garbled output when `DEBUG_MAC` is activated for example.

Ah yes, good point. I was using this for troubleshooting reasons and did not think of removing it for this pull request. I just removed the flags.

User-specific flags can be added to `CONFIG.mine`, which will be copied into the image at built time.

As an aside, there are many ways to write the `Dockerfile`. We've been using different variations. This proposed variation works well for us right now but If you have suggestions, please let me know, I'll be happy to make necessary modifications.

Also, I could add a small example of how to build the image. I'm not sure where to put this ... perhaps in the `README.md` as an additional **TL;DR (Docker)** section?

sbellem commented:
> Do you think setting 256 as the default bit length is specific to your application or of general interest? 128 bits makes sense to me because it allows 64-bit computation with 40-bit security. `GFP_MOD_SZ` isn't as important as it used to be, but it might confuse people looking at the code.

Ah yes, it's specific to our application. I will change this.

sbellem commented:
@mkskeller I've modified the `Dockerfile` a bit such that by default it builds the `mascot-party.x` with 128 bit length (`GFP_MOD_SZ=2`). I have left the build argument `GFP_MOD_SZ` there because it can be convenient when building for 256 bit length. Similarly there are a few other build arguments in the `Dockerfile` which are there for convenience such that one does not need to add or edit the `CONFIG.mine` when building.

The build arguments, included for convenience, but which can all be set in `CONFIG.mine`, are:

```dockerfile
ARG cxx=clang++-11
ARG use_ntl=0
ARG prep_dir="Player-Data"
ARG ssl_dir="Player-Data"
ARG gfp_mod_sz=2
```

For the `compile.py` phase and for the ssl key setup, the following arguments are provided:

```dockerfile
ARG cryptoplayers=0
ARG compile_options="--field=64"
ARG src="tutorial"
```

Since some scripts (e.g. `mal-shamir.sh`) rely on the `PLAYERS` environment variable, it's set in:

```dockerfile
ENV PLAYERS ${cryptoplayers}
```

Again, this is just for convenience, and ultimately, someone wishing to build and run different combinations can create a `CONFIG.mine` file,  and will work in a container.

The `Dockerfile` is structured in 3 stages:

1. Build environment
2. Virtual Machine build
3. High-level program compilation

I've added comments in the `Dockerfile` that explain how to work with each stage. I was not sure where to put this documentation so I included in the `Dockerfile` itself for now.

mkskeller commented:
Looks very good now, thank you!

sbellem commented:
Thank you!
